File: QLCTApp/BankDirectory.swift
import Foundation

enum BankCategory: String, Codable {
    case domesticCommercial
    case transformedBank
    case jointVenture
    case foreignOwned
    case policyBank
    case cooperativeBank
    case custom
}

struct BankDefinition: Identifiable, Codable, Hashable {
    let id: String
    let displayName: String
    let shortName: String
    let legalName: String
    let category: BankCategory
    let aliases: [String]
    let legacyAliases: [String]
    let appHints: [String]
    let supportsGenericParser: Bool
    let isActive: Bool
}

enum BankMatchType: String, Codable {
    case exact
    case alias
    case legacyAlias
    case appHint
    case fuzzy
    case manual
}

struct BankMatchResult {
    let bank: BankDefinition
    let confidence: Double
    let matchedAlias: String
    let matchType: BankMatchType
}

enum BankTextNormalizer {
    static func normalize(_ text: String) -> String {
        text
            .folding(options: [.diacriticInsensitive, .caseInsensitive], locale: Locale(identifier: "vi_VN"))
            .uppercased()
            .replacingOccurrences(of: #"[-_:/().,;]"#, with: " ", options: .regularExpression)
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

enum BankMatcher {
    static func match(_ text: String, banks: [BankDefinition] = VietnamBankDirectory.banks) -> BankMatchResult? {
        let normalizedText = BankTextNormalizer.normalize(text)
        var results: [(BankMatchResult, Int)] = []

        for bank in banks where bank.isActive {
            let candidates: [(String, BankMatchType, Double)] =
                [(bank.displayName, .exact, 1.0), (bank.shortName, .exact, 1.0)]
                + bank.aliases.map { ($0, .alias, 0.95) }
                + bank.legacyAliases.map { ($0, .legacyAlias, 0.90) }
                + bank.appHints.map { ($0, .appHint, 0.85) }

            for candidate in candidates {
                let alias = BankTextNormalizer.normalize(candidate.0)
                guard !alias.isEmpty, containsToken(alias, in: normalizedText) else { continue }
                results.append((
                    BankMatchResult(bank: bank, confidence: candidate.2, matchedAlias: candidate.0, matchType: candidate.1),
                    alias.count
                ))
            }
        }

        return results
            .sorted { left, right in
                if left.1 != right.1 { return left.1 > right.1 }
                return left.0.confidence > right.0.confidence
            }
            .first?.0
    }

    private static func containsToken(_ alias: String, in text: String) -> Bool {
        let escaped = NSRegularExpression.escapedPattern(for: alias)
        let pattern = "(^|\\s)\(escaped)(?=\\s|$)"
        return text.range(of: pattern, options: .regularExpression) != nil
    }
}

enum VietnamBankDirectory {
    static let banks: [BankDefinition] = [
        bank("agribank", "Agribank", "Agribank", "Ngân hàng Nông nghiệp và Phát triển Nông thôn Việt Nam", .domesticCommercial, ["Agribank", "VBARD", "NHNo", "Ngân hàng Nông nghiệp"], [], ["Agribank Plus"]),
        bank("acb", "ACB", "ACB", "Ngân hàng TMCP Á Châu", .domesticCommercial, ["ACB", "ACB Bank", "Asia Commercial Bank", "Ngân hàng Á Châu"], [], ["ACB ONE"]),
        bank("abbank", "ABBANK", "ABBANK", "Ngân hàng TMCP An Bình", .domesticCommercial, ["ABBANK", "AB Bank", "An Binh Bank"], [], ["ABBANK"]),
        bank("bac_a_bank", "Bac A Bank", "BAB", "Ngân hàng TMCP Bắc Á", .domesticCommercial, ["BAC A BANK", "BacABank", "BACA BANK", "Ngân hàng Bắc Á"], [], ["Bac A Bank"]),
        bank("bvbank", "BVBank", "BVB", "Ngân hàng TMCP Bản Việt", .domesticCommercial, ["BVBANK", "BV Bank", "Ngân hàng Bản Việt"], ["Viet Capital Bank", "VietCapital Bank"], ["digimi", "DigiMi"]),
        bank("baoviet_bank", "BAOVIET Bank", "BAOVIETBANK", "Ngân hàng TMCP Bảo Việt", .domesticCommercial, ["BAOVIET BANK", "Bao Viet Bank", "BAOVIETBANK"], [], ["BAOVIET Bank"]),
        bank("vietinbank", "VietinBank", "VietinBank", "Ngân hàng TMCP Công Thương Việt Nam", .domesticCommercial, ["VIETINBANK", "Vietin Bank", "Ngân hàng Công Thương"], ["Incombank", "ICBV"], ["VietinBank iPay"]),
        bank("pvcombank", "PVcomBank", "PVcomBank", "Ngân hàng TMCP Đại Chúng Việt Nam", .domesticCommercial, ["PVCOMBANK", "PV COM BANK", "PVcom Bank"], ["Western Bank", "PVFC"], ["PVcomBank"]),
        bank("bidv", "BIDV", "BIDV", "Ngân hàng TMCP Đầu tư và Phát triển Việt Nam", .domesticCommercial, ["BIDV", "Bank for Investment and Development of Vietnam"], [], ["BIDV SmartBanking"]),
        bank("seabank", "SeABank", "SeABank", "Ngân hàng TMCP Đông Nam Á", .domesticCommercial, ["SEABANK", "SEA BANK", "Ngân hàng Đông Nam Á"], [], ["SeAMobile"]),
        bank("msb", "MSB", "MSB", "Ngân hàng TMCP Hàng Hải Việt Nam", .domesticCommercial, ["MSB", "MSB Bank"], ["Maritime Bank", "MaritimeBank", "Vietnam Maritime Bank"], ["MSB mBank"]),
        bank("kienlongbank", "KienlongBank", "KLB", "Ngân hàng TMCP Kiên Long", .domesticCommercial, ["KIENLONGBANK", "Kien Long Bank", "KLB"], [], ["KienlongBank"]),
        bank("techcombank", "Techcombank", "TCB", "Ngân hàng TMCP Kỹ Thương Việt Nam", .domesticCommercial, ["TECHCOMBANK", "Techcom Bank", "TCB"], [], ["Techcombank Mobile"]),
        bank("lpbank", "LPBank", "LPBank", "Ngân hàng TMCP Lộc Phát Việt Nam", .domesticCommercial, ["LPBANK", "LP BANK"], ["LienVietPostBank", "Lien Viet Post Bank", "LienVietBank", "LVPB"], ["LPBank"]),
        bank("nam_a_bank", "Nam A Bank", "NAB", "Ngân hàng TMCP Nam Á", .domesticCommercial, ["NAM A BANK", "NamABank", "NAB"], [], ["Nam A Bank"]),
        bank("vietcombank", "Vietcombank", "VCB", "Ngân hàng TMCP Ngoại thương Việt Nam", .domesticCommercial, ["VIETCOMBANK", "Vietcom Bank", "VCB", "VCB Digibank"], [], ["Vietcombank Digibank"]),
        bank("hdbank", "HDBank", "HDBank", "Ngân hàng TMCP Phát triển TP.HCM", .domesticCommercial, ["HDBANK", "HD BANK"], ["Housing Development Bank"], ["HDBank"]),
        bank("ocb", "OCB", "OCB", "Ngân hàng TMCP Phương Đông", .domesticCommercial, ["OCB", "OCB Bank", "Orient Commercial Bank"], ["Orient Bank"], ["OCB OMNI"]),
        bank("mbbank", "MB", "MB", "Ngân hàng TMCP Quân Đội", .domesticCommercial, ["MBBANK", "MB Bank", "Military Bank", "Ngân hàng Quân Đội"], [], ["MBBank"]),
        bank("ncb", "NCB", "NCB", "Ngân hàng TMCP Quốc Dân", .domesticCommercial, ["NCB", "NCB Bank", "National Citizen Bank"], ["Navibank", "NaviBank"], ["NCB"]),
        bank("vib", "VIB", "VIB", "Ngân hàng TMCP Quốc tế Việt Nam", .domesticCommercial, ["VIB", "VIB Bank", "Vietnam International Bank"], [], ["MyVIB"]),
        bank("scb_vietnam", "SCB", "SCB", "Ngân hàng TMCP Sài Gòn", .domesticCommercial, ["SCB", "SCB Vietnam", "Saigon Commercial Bank"], [], ["SCB"]),
        bank("shb", "SHB", "SHB", "Ngân hàng TMCP Sài Gòn - Hà Nội", .domesticCommercial, ["SHB", "SHB Bank", "Saigon Hanoi Bank"], [], ["SHB Mobile"]),
        bank("saigonbank", "SAIGONBANK", "SGB", "Ngân hàng TMCP Sài Gòn Công Thương", .domesticCommercial, ["SAIGONBANK", "Saigon Bank", "SGB"], [], ["SAIGONBANK"]),
        bank("sacombank", "Sacombank", "Sacombank", "Ngân hàng TMCP Sài Gòn Tài Lộc", .domesticCommercial, ["SACOMBANK", "Sacom Bank", "Ngân hàng Sài Gòn Tài Lộc"], ["STB", "Saigon Thuong Tin", "Sài Gòn Thương Tín"], ["Sacombank Pay"]),
        bank("pgbank", "PGBank", "PGBank", "Ngân hàng TMCP Thịnh Vượng và Phát triển", .domesticCommercial, ["PGBANK", "PG Bank"], ["Petrolimex Bank"], ["PGBank"]),
        bank("tpbank", "TPBank", "TPBank", "Ngân hàng TMCP Tiên Phong", .domesticCommercial, ["TPBANK", "TP Bank", "Tien Phong Bank", "Tiên Phong"], [], ["TPBank"]),
        bank("vietabank", "VietABank", "VAB", "Ngân hàng TMCP Việt Á", .domesticCommercial, ["VIETABANK", "Viet A Bank", "VAB"], [], ["VietABank"]),
        bank("vpbank", "VPBank", "VPBank", "Ngân hàng TMCP Việt Nam Thịnh Vượng", .domesticCommercial, ["VPBANK", "VP Bank", "VPB"], [], ["VPBank NEO"]),
        bank("vietbank", "Vietbank", "Vietbank", "Ngân hàng TMCP Việt Nam Thương Tín", .domesticCommercial, ["VIETBANK", "Viet Bank"], [], ["Vietbank"]),
        bank("eximbank", "Eximbank", "EIB", "Ngân hàng TMCP Xuất Nhập khẩu Việt Nam", .domesticCommercial, ["EXIMBANK", "Exim Bank", "EIB"], [], ["Eximbank"]),
        bank("gpbank", "GPBank", "GPBank", "Ngân hàng Thương mại TNHH MTV Kỷ Nguyên Thịnh Vượng", .transformedBank, ["GPBANK", "GP Bank"], [], ["GPBank"]),
        bank("vcbneo", "VCBNeo", "VCBNeo", "Ngân hàng Thương mại TNHH MTV Ngoại thương Công nghệ số", .transformedBank, ["VCBNEO", "VCB Neo", "CBBank", "CB Bank"], [], ["VCBNeo"]),
        bank("vikki_bank", "Vikki Bank", "Vikki", "Ngân hàng TNHH MTV Số Vikki", .transformedBank, ["VIKKI BANK", "Vikki", "Vikki DAB"], ["DongA Bank", "Dong A Bank", "DAB"], ["Vikki Bank"]),
        bank("mbv", "MBV", "MBV", "Ngân hàng TNHH MTV Việt Nam Hiện Đại", .transformedBank, ["MBV"], ["OceanBank", "Ocean Bank"], ["MBV"]),
        bank("ivb", "Indovina Bank", "IVB", "Ngân hàng TNHH Indovina", .jointVenture, ["Indovina Bank", "IVB"], [], ["IVB"]),
        bank("vrb", "VRB", "VRB", "Ngân hàng Liên doanh Việt - Nga", .jointVenture, ["VRB", "Vietnam Russia Bank"], [], ["VRB"]),
        bank("anz_vietnam", "ANZ Vietnam", "ANZ", "ANZ Vietnam", .foreignOwned, ["ANZ", "ANZ Vietnam"], [], ["ANZ"]),
        bank("cimb_vietnam", "CIMB Vietnam", "CIMB", "CIMB Bank Vietnam Limited", .foreignOwned, ["CIMB", "CIMB Bank", "CIMB Vietnam"], [], ["CIMB"]),
        bank("hong_leong_vietnam", "Hong Leong Bank Vietnam", "HLB", "Hong Leong Bank Vietnam Limited", .foreignOwned, ["Hong Leong Bank", "Hong Leong Vietnam", "HLB"], [], ["Hong Leong Bank"]),
        bank("hsbc_vietnam", "HSBC Vietnam", "HSBC", "HSBC Bank Vietnam Limited", .foreignOwned, ["HSBC", "HSBC Bank", "HSBC Vietnam"], [], ["HSBC"]),
        bank("public_bank_vietnam", "Public Bank Vietnam", "PBVN", "Public Bank Vietnam Limited", .foreignOwned, ["Public Bank", "Public Bank Vietnam", "PBVN"], ["VID Public Bank"], ["Public Bank"]),
        bank("shinhan_vietnam", "Shinhan Bank Vietnam", "Shinhan", "Shinhan Bank Vietnam Limited", .foreignOwned, ["Shinhan", "Shinhan Bank", "Shinhan Vietnam"], [], ["Shinhan SOL"]),
        bank("standard_chartered_vietnam", "Standard Chartered Vietnam", "Standard Chartered", "Standard Chartered Bank Vietnam Limited", .foreignOwned, ["Standard Chartered", "Standard Chartered Bank"], [], ["Standard Chartered"]),
        bank("uob_vietnam", "UOB Vietnam", "UOB", "United Overseas Bank Vietnam Limited", .foreignOwned, ["UOB", "UOB Bank", "UOB Vietnam"], [], ["UOB TMRW"]),
        bank("woori_vietnam", "Woori Bank Vietnam", "Woori", "Woori Bank Vietnam Limited", .foreignOwned, ["Woori", "Woori Bank", "Woori Vietnam"], [], ["Woori WON"]),
        bank("vbsp", "Ngân hàng Chính sách Xã hội", "VBSP", "Ngân hàng Chính sách Xã hội Việt Nam", .policyBank, ["VBSP", "NHCSXH", "Vietnam Bank for Social Policies"], [], ["VBSP"]),
        bank("vdb", "Ngân hàng Phát triển Việt Nam", "VDB", "Ngân hàng Phát triển Việt Nam", .policyBank, ["VDB", "Vietnam Development Bank"], [], ["VDB"]),
        bank("coopbank", "Co-opBank", "Co-opBank", "Ngân hàng Hợp tác xã Việt Nam", .cooperativeBank, ["COOPBANK", "CO-OPBANK", "Coop Bank", "Cooperative Bank of Vietnam"], [], ["Co-opBank"])
    ]

    private static func bank(
        _ id: String,
        _ displayName: String,
        _ shortName: String,
        _ legalName: String,
        _ category: BankCategory,
        _ aliases: [String],
        _ legacyAliases: [String],
        _ appHints: [String]
    ) -> BankDefinition {
        BankDefinition(
            id: id,
            displayName: displayName,
            shortName: shortName,
            legalName: legalName,
            category: category,
            aliases: aliases,
            legacyAliases: legacyAliases,
            appHints: appHints,
            supportsGenericParser: true,
            isActive: true
        )
    }
}

File: bank-directory.js
const VIETNAM_BANKS = [
  ["agribank", "Agribank", "Agribank", "domesticCommercial", "#b61f2c", ["agribank", "vietnam bank for agriculture", "vbard", "nhno", "nhno pttn"]],
  ["acb", "ACB", "ACB", "domesticCommercial", "#1269b0", ["acb", "acb bank", "asia commercial bank", "ngan hang a chau"]],
  ["abbank", "ABBANK", "AB", "domesticCommercial", "#19a15f", ["abbank", "ab bank", "an binh bank"]],
  ["bac_a_bank", "Bac A Bank", "BAB", "domesticCommercial", "#0f7d4f", ["bac a bank", "bacabank", "baca bank", "ngan hang bac a"]],
  ["bvbank", "BVBank", "BVB", "domesticCommercial", "#1f8ed5", ["bvbank", "bv bank", "viet capital bank", "vietcapital bank", "ngan hang ban viet"]],
  ["baoviet_bank", "BAOVIET Bank", "BV", "domesticCommercial", "#f0a000", ["baoviet bank", "bao viet bank", "baovietbank"]],
  ["vietinbank", "VietinBank", "VTB", "domesticCommercial", "#0f62a8", ["vietinbank", "vietin bank", "incombank", "icbv", "ngan hang cong thuong"]],
  ["pvcombank", "PVcomBank", "PVC", "domesticCommercial", "#008fb3", ["pvcombank", "pv com bank", "pvcom bank", "western bank", "pvfc"]],
  ["bidv", "BIDV", "BIDV", "domesticCommercial", "#0069aa", ["bidv", "bank for investment and development", "ngan hang dau tu va phat trien"]],
  ["seabank", "SeABank", "SEA", "domesticCommercial", "#e45a20", ["seabank", "sea bank", "seamobile", "ngan hang dong nam a"]],
  ["msb", "MSB", "MSB", "domesticCommercial", "#f05a22", ["msb", "msb bank", "maritime bank", "maritimebank", "vietnam maritime bank"]],
  ["kienlongbank", "KienlongBank", "KLB", "domesticCommercial", "#4d9b37", ["kienlongbank", "kien long bank", "kienlong bank", "klb"]],
  ["techcombank", "Techcombank", "TCB", "domesticCommercial", "#e31b23", ["techcombank", "techcom bank", "tcb"]],
  ["lpbank", "LPBank", "LPB", "domesticCommercial", "#8b1d5b", ["lpbank", "lp bank", "lienvietpostbank", "lien viet post bank", "lienvietbank", "lvpb"]],
  ["nam_a_bank", "Nam A Bank", "NAB", "domesticCommercial", "#0f78bd", ["nam a bank", "namabank", "nam a", "nab"]],
  ["vietcombank", "Vietcombank", "VCB", "domesticCommercial", "#007a3d", ["vietcombank", "vietcom bank", "vcb digibank", "vcb", "bank for foreign trade"]],
  ["hdbank", "HDBank", "HDB", "domesticCommercial", "#d91f26", ["hdbank", "hd bank", "housing development bank"]],
  ["ocb", "OCB", "OCB", "domesticCommercial", "#18a05e", ["ocb", "ocb bank", "ocb omni", "orient commercial bank"]],
  ["mbbank", "MB", "MB", "domesticCommercial", "#1b4f9c", ["mbbank", "mb bank", "military bank", "ngan hang quan doi"]],
  ["ncb", "NCB", "NCB", "domesticCommercial", "#c49a38", ["ncb", "ncb bank", "national citizen bank", "navibank", "navi bank"]],
  ["vib", "VIB", "VIB", "domesticCommercial", "#f58220", ["vib", "vib bank", "myvib", "vietnam international bank"]],
  ["scb_vietnam", "SCB", "SCB", "domesticCommercial", "#113f8c", ["scb", "scb vietnam", "saigon commercial bank"]],
  ["shb", "SHB", "SHB", "domesticCommercial", "#f58220", ["shb", "shb bank", "saigon hanoi bank", "saigon-hanoi bank"]],
  ["saigonbank", "SAIGONBANK", "SGB", "domesticCommercial", "#0072bc", ["saigonbank", "saigon bank", "sgb"]],
  ["sacombank", "Sacombank", "STB", "domesticCommercial", "#005baa", ["sacombank", "sacom bank", "stb", "saigon thuong tin", "sai gon thuong tin", "ngan hang sai gon tai loc"]],
  ["pgbank", "PGBank", "PGB", "domesticCommercial", "#f47b20", ["pgbank", "pg bank", "petrolimex bank", "pg bank petrolimex"]],
  ["tpbank", "TPBank", "TPB", "domesticCommercial", "#5c2483", ["tpbank", "tp bank", "tien phong bank", "tienphong bank"]],
  ["vietabank", "VietABank", "VAB", "domesticCommercial", "#d71920", ["vietabank", "viet a bank", "vab", "ngan hang viet a"]],
  ["vpbank", "VPBank", "VPB", "domesticCommercial", "#1f9c4d", ["vpbank", "vp bank", "vpb", "vietnam prosperity bank"]],
  ["vietbank", "Vietbank", "VBB", "domesticCommercial", "#005bac", ["vietbank", "viet bank", "viet nam thuong tin bank"]],
  ["eximbank", "Eximbank", "EIB", "domesticCommercial", "#0b78bd", ["eximbank", "exim bank", "eib"]],
  ["gpbank", "GPBank", "GPB", "transformedBank", "#009b72", ["gpbank", "gp bank", "global petro bank"]],
  ["vcbneo", "VCBNeo", "NEO", "transformedBank", "#00a859", ["vcbneo", "vcb neo", "cb bank", "cbbank", "construction bank"]],
  ["vikki_bank", "Vikki Bank", "VIK", "transformedBank", "#ef476f", ["vikki bank", "vikki", "dong a bank", "donga bank", "dab", "vikki dab"]],
  ["mbv", "MBV", "MBV", "transformedBank", "#1b4f9c", ["mbv", "modern bank of vietnam", "oceanbank", "ocean bank"]],
  ["ivb", "Indovina Bank", "IVB", "jointVenture", "#006b54", ["indovina bank", "ivb", "indovina"]],
  ["vrb", "VRB", "VRB", "jointVenture", "#0055a4", ["vrb", "viet nga bank", "vietnam russia bank", "vietnam-russia bank"]],
  ["anz_vietnam", "ANZ Vietnam", "ANZ", "foreignOwned", "#0073cf", ["anz", "anz vietnam", "anz bank vietnam"]],
  ["cimb_vietnam", "CIMB Vietnam", "CIMB", "foreignOwned", "#b11f2a", ["cimb", "cimb bank", "cimb vietnam"]],
  ["hong_leong_vietnam", "Hong Leong Bank Vietnam", "HLB", "foreignOwned", "#d71920", ["hong leong bank", "hong leong vietnam", "hlb"]],
  ["hsbc_vietnam", "HSBC Vietnam", "HSBC", "foreignOwned", "#db0011", ["hsbc", "hsbc bank", "hsbc vietnam", "hongkong and shanghai banking corporation"]],
  ["public_bank_vietnam", "Public Bank Vietnam", "PBVN", "foreignOwned", "#e31b23", ["public bank", "public bank vietnam", "pbvn", "vid public bank"]],
  ["shinhan_vietnam", "Shinhan Bank Vietnam", "SHB", "foreignOwned", "#0046ad", ["shinhan", "shinhan bank", "shinhan sol", "sol"]],
  ["standard_chartered_vietnam", "Standard Chartered Vietnam", "SC", "foreignOwned", "#0f8f61", ["standard chartered", "standard chartered bank", "standard chartered vietnam"]],
  ["uob_vietnam", "UOB Vietnam", "UOB", "foreignOwned", "#004b9b", ["uob", "uob bank", "uob vietnam", "uob tmrw", "tmrw"]],
  ["woori_vietnam", "Woori Bank Vietnam", "WOO", "foreignOwned", "#0067b1", ["woori", "woori bank", "woori vietnam", "woori won"]],
  ["vbsp", "Ngân hàng Chính sách Xã hội", "VBSP", "policyBank", "#007a3d", ["vbsp", "nhcsxh", "vietnam bank for social policies", "ngan hang chinh sach xa hoi"]],
  ["vdb", "Ngân hàng Phát triển Việt Nam", "VDB", "policyBank", "#0069aa", ["vdb", "vietnam development bank", "ngan hang phat trien viet nam"]],
  ["coopbank", "Co-opBank", "COOP", "cooperativeBank", "#00845f", ["coopbank", "co-opbank", "coop bank", "cooperative bank of vietnam", "cov"]]
].map(([id, displayName, shortName, category, color, aliases]) => ({
  id,
  displayName,
  shortName,
  category,
  color,
  aliases
}))

const BANK_LOGO_OVERRIDES = {
  anz_vietnam: "assets/banks/anz_vietnam.svg"
}

const BANK_LOGO_MISSING = new Set(["vdb"])

function normalizeBankText(value) {
  return String(value || "")
    .normalize("NFD")
    .replace(/\p{Diacritic}/gu, "")
    .toUpperCase()
    .replace(/[-_:/().,;]/g, " ")
    .replace(/\s+/g, " ")
    .trim()
}

function aliasPattern(alias) {
  const normalized = normalizeBankText(alias).replace(/[.*+?^${}()|[\]\\]/g, "\\$&")
  return new RegExp(`(^|\\s)${normalized}(?=\\s|$)`, "i")
}

function findBankByID(id) {
  return VIETNAM_BANKS.find(bank => bank.id === id)
}

function findBankByName(name) {
  const normalized = normalizeBankText(name)
  return VIETNAM_BANKS.find(bank =>
    normalizeBankText(bank.displayName) === normalized ||
    normalizeBankText(bank.shortName) === normalized
  )
}

function matchBank(text) {
  const normalized = normalizeBankText(text)
  const matches = []

  for (const bank of VIETNAM_BANKS) {
    const aliases = [bank.displayName, bank.shortName, ...bank.aliases]
    for (const alias of aliases) {
      const normalizedAlias = normalizeBankText(alias)
      if (!normalizedAlias) continue
      if (aliasPattern(alias).test(normalized)) {
        matches.push({
          bank,
          matchedAlias: alias,
          confidence: normalizedAlias === normalizeBankText(bank.displayName) || normalizedAlias === normalizeBankText(bank.shortName) ? 1 : 0.95,
          length: normalizedAlias.length
        })
      }
    }
  }

  return matches.sort((a, b) => b.length - a.length || b.confidence - a.confidence)[0] || null
}

function bankLogo(bankOrName, size = "regular") {
  const bank = typeof bankOrName === "string"
    ? findBankByID(bankOrName) || findBankByName(bankOrName) || matchBank(bankOrName)?.bank
    : bankOrName

  const color = bank?.color || "#00899a"
  const shortName = bank?.shortName || "BANK"
  const label = shortName.length > 4 ? shortName.slice(0, 4) : shortName
  const logoPath = bank && !BANK_LOGO_MISSING.has(bank.id)
    ? BANK_LOGO_OVERRIDES[bank.id] || `assets/banks/${bank.id}.png`
    : ""

  return `
    <div class="bank-logo ${size} ${logoPath ? "has-image" : ""}" style="--bank-color:${color}" title="${bank?.displayName || "Ngân hàng"}">
      ${logoPath ? `<img src="${logoPath}" alt="${bank.displayName}" loading="lazy" onerror="this.hidden=true;this.parentElement.classList.remove('has-image')" />` : ""}
      <span>${label}</span>
    </div>
  `
}

function iconSvg(name) {
  const icons = {
    home: '<svg viewBox="0 0 24 24"><path d="m3 11 9-7 9 7v9a1 1 0 0 1-1 1h-5v-6H9v6H4a1 1 0 0 1-1-1v-9Z"/></svg>',
    list: '<svg viewBox="0 0 24 24"><path d="M8 6h12M8 12h12M8 18h12"/><path d="M4 6h.01M4 12h.01M4 18h.01"/></svg>',
    clock: '<svg viewBox="0 0 24 24"><path d="M12 21a9 9 0 1 0 0-18 9 9 0 0 0 0 18Z"/><path d="M12 7v5l3 2"/></svg>',
    bell: '<svg viewBox="0 0 24 24"><path d="M6 9a6 6 0 1 1 12 0c0 7 3 7 3 9H3c0-2 3-2 3-9Z"/><path d="M10 21h4"/></svg>',
    shield: '<svg viewBox="0 0 24 24"><path d="M12 3 20 6v6c0 5-3.4 8-8 9-4.6-1-8-4-8-9V6l8-3Z"/><path d="m9 12 2 2 4-5"/></svg>',
    key: '<svg viewBox="0 0 24 24"><path d="M15 7a4 4 0 1 0 2 3.5L22 15v3h-3v3h-3l-4.5-4.5"/><path d="M7 10h.01"/></svg>',
    upload: '<svg viewBox="0 0 24 24"><path d="M12 16V4"/><path d="m7 9 5-5 5 5"/><path d="M5 20h14"/></svg>',
    download: '<svg viewBox="0 0 24 24"><path d="M12 4v12"/><path d="m7 11 5 5 5-5"/><path d="M5 20h14"/></svg>',
    user: '<svg viewBox="0 0 24 24"><path d="M12 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8Z"/><path d="M4 21a8 8 0 0 1 16 0"/></svg>',
    refresh: '<svg viewBox="0 0 24 24"><path d="M20 7v5h-5"/><path d="M4 17v-5h5"/><path d="M6 9a7 7 0 0 1 11.7-2.7L20 8"/><path d="M18 15a7 7 0 0 1-11.7 2.7L4 16"/></svg>',
    trash: '<svg viewBox="0 0 24 24"><path d="M4 7h16"/><path d="M10 11v6M14 11v6"/><path d="M6 7l1 14h10l1-14"/><path d="M9 7V4h6v3"/></svg>',
    info: '<svg viewBox="0 0 24 24"><path d="M12 21a9 9 0 1 0 0-18 9 9 0 0 0 0 18Z"/><path d="M12 11v6M12 7h.01"/></svg>',
    coin: '<svg viewBox="0 0 24 24"><path d="M12 20c5 0 9-2 9-4.5V8.5C21 6 17 4 12 4S3 6 3 8.5v7C3 18 7 20 12 20Z"/><path d="M3 8.5C3 11 7 13 12 13s9-2 9-4.5"/><path d="M3 12c0 2.5 4 4.5 9 4.5s9-2 9-4.5"/></svg>',
    wallet: '<svg viewBox="0 0 24 24"><path d="M4 7.5A2.5 2.5 0 0 1 6.5 5H18a2 2 0 0 1 2 2v2h-4.5a3.5 3.5 0 0 0 0 7H20v1a2 2 0 0 1-2 2H6.5A2.5 2.5 0 0 1 4 16.5v-9Z"/><path d="M15.5 11H21v3h-5.5a1.5 1.5 0 0 1 0-3Z"/></svg>',
    receipt: '<svg viewBox="0 0 24 24"><path d="M7 3h10a1 1 0 0 1 1 1v17l-3-2-3 2-3-2-3 2V4a1 1 0 0 1 1-1Z"/><path d="M9 8h6M9 12h6M9 16h4"/></svg>',
    piggy: '<svg viewBox="0 0 24 24"><path d="M5 12a6 6 0 0 1 6-6h3.5A4.5 4.5 0 0 1 19 10.5h1.5V15H19a5 5 0 0 1-2 2.3V20h-3v-2h-4v2H7v-2.4A6 6 0 0 1 5 12Z"/><path d="M8 8 6 5M15 10h.01"/></svg>',
    card: '<svg viewBox="0 0 24 24"><path d="M4 6h16a2 2 0 0 1 2 2v8a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2Z"/><path d="M2 10h20M6 15h5"/></svg>',
    house: '<svg viewBox="0 0 24 24"><path d="m3 11 9-7 9 7v9a1 1 0 0 1-1 1h-5v-6H9v6H4a1 1 0 0 1-1-1v-9Z"/></svg>',
    laptop: '<svg viewBox="0 0 24 24"><path d="M6 5h12a1 1 0 0 1 1 1v9H5V6a1 1 0 0 1 1-1Z"/><path d="M3 18h18"/></svg>',
    bolt: '<svg viewBox="0 0 24 24"><path d="m13 2-8 12h6l-1 8 9-13h-6l0-7Z"/></svg>',
    wifi: '<svg viewBox="0 0 24 24"><path d="M4 10a12 12 0 0 1 16 0M7 13a7 7 0 0 1 10 0M10 16a3 3 0 0 1 4 0"/><path d="M12 20h.01"/></svg>',
    bank: '<svg viewBox="0 0 24 24"><path d="M3 10h18L12 4 3 10Z"/><path d="M5 10v8M9 10v8M15 10v8M19 10v8M4 20h16"/></svg>',
    check: '<svg viewBox="0 0 24 24"><path d="m5 12 4 4L19 6"/></svg>',
    calendar: '<svg viewBox="0 0 24 24"><path d="M7 3v4M17 3v4M4 8h16M5 5h14a1 1 0 0 1 1 1v14H4V6a1 1 0 0 1 1-1Z"/></svg>',
    chart: '<svg viewBox="0 0 24 24"><path d="M4 19V5M4 19h16"/><path d="M8 16v-5M12 16V8M16 16v-9"/></svg>',
    settings: '<svg viewBox="0 0 24 24"><path d="M12 15.5A3.5 3.5 0 1 0 12 8a3.5 3.5 0 0 0 0 7.5Z"/><path d="m19 13.5 2-1.5-2-1.5-.5-1.3.8-2.4-2.5-1.4-1.8 1.7-1.4-.2L12 4h-3l-.7 2.9-1.4.2L5.1 5.4 2.6 6.8l.8 2.4-.5 1.3L1 12l1.9 1.5.5 1.3-.8 2.4 2.5 1.4 1.8-1.7 1.4.2L9 20h3l.7-2.9 1.4-.2 1.8 1.7 2.5-1.4-.8-2.4.4-1.3Z"/></svg>'
  }

  return icons[name] || icons.card
}
